drivers/flash/spi_nor.c

```diff
@@ -785,9 +919,121 @@ static int mxicy_configure(const struct device *dev, const uint8_t *jedec_id)
+ * function.
+ *
+ * @param dev Device struct
+ * @param sr The new value of the configuration register2
+ *
+ * @return 0 on success or a negative error code.
+ */
+static int spi_nor_wrcr2(const struct device *dev,
+			 uint8_t cr)
+{
+	int ret = spi_nor_cmd_write(dev, SPI_NOR_CMD_WREN);
+
+	if (ret == 0) {
+
+		struct spi_nor_data *data = dev->data;
+
+		data->flag_access_32bit = true;
+
+		ret = spi_nor_cmd_addr_write(dev, SPI_NOR_CMD_WR_CFGREG2, 0, &cr, sizeof(cr));
+	}
+
+	return ret;
+}
+
+static int spi_nor_change_protocol(const struct device *dev,
+						enum spi_nor_protocol nor_protocol)
+{
+	struct spi_nor_data *data = dev->data;
+	uint8_t cr2;
+	int ret = 0;
+
+	switch (nor_protocol) {
+	case PROTO_1_1_1:
+		data->nor_protocol = PROTO_1_1_1;
+		data->flag_access_32bit = false;
+		data->read_cmd = SPI_NOR_CMD_READ;
+		data->program_cmd = SPI_NOR_CMD_PP;
+		data->erase_cmd = SPI_NOR_CMD_SE;
+		break;
+	case PROTO_1_1_2:
+		data->nor_protocol = PROTO_1_1_2;
+		data->flag_access_32bit = false;
+		data->read_cmd = SPI_NOR_CMD_READ;
+		data->program_cmd = SPI_NOR_CMD_PP;
+		data->erase_cmd = SPI_NOR_CMD_SE;
+		break;
+	case PROTO_1_2_2:
+		data->nor_protocol = PROTO_1_2_2;
+		data->flag_access_32bit = false;
+		data->read_cmd = SPI_NOR_CMD_READ;
+		data->program_cmd = SPI_NOR_CMD_PP;
+		data->erase_cmd = SPI_NOR_CMD_SE;
+		break;
+	case PROTO_1_1_4:
+		data->nor_protocol = PROTO_1_1_4;
+		data->flag_access_32bit = false;
+		data->read_cmd = SPI_NOR_CMD_READ;
+		data->program_cmd = SPI_NOR_CMD_PP;
+		data->erase_cmd = SPI_NOR_CMD_SE;
+		break;
+	case PROTO_1_4_4:
+		data->nor_protocol = PROTO_1_4_4;
+		data->flag_access_32bit = false;
+		data->read_cmd = SPI_NOR_CMD_READ;
+		data->program_cmd = SPI_NOR_CMD_PP;
+		data->erase_cmd = SPI_NOR_CMD_SE;
+		break;
+	case PROTO_1_4D_4D:
+		data->nor_protocol = PROTO_1_4D_4D;
+		data->flag_access_32bit = false;
+		data->read_cmd = SPI_NOR_CMD_READ;
+		data->program_cmd = SPI_NOR_CMD_PP;
+		data->erase_cmd = SPI_NOR_CMD_SE;
+		break;
+	case PROTO_4_4_4:
+		ret = spi_nor_cmd_write(dev, SPI_NOR_CMD_EQIO);
+		data->nor_protocol = PROTO_4_4_4;
+		data->flag_access_32bit = false;
+		data->read_cmd = SPI_NOR_CMD_4READ;
+		data->program_cmd = SPI_NOR_CMD_PP;
+		data->erase_cmd = SPI_NOR_CMD_SE;
+		break;
+	case PROTO_8_8_8:
+		data->flag_access_32bit = true;
+		data->read_cmd = SPI_NOR_OCMD_RD;
+		data->program_cmd = SPI_NOR_CMD_PP_4B;
+		data->erase_cmd = SPI_NOR_OCMD_SE;
+		cr2 = 0x01;
+		ret = spi_nor_wrcr2(dev, cr2);
+		data->nor_protocol = PROTO_8_8_8;
+		break;
+	case PROTO_8D_8D_8D:
+		data->flag_access_32bit = true;
+		data->read_cmd = SPI_NOR_OCMD_DTR_RD;
+		data->program_cmd = SPI_NOR_CMD_PP_4B;
+		data->erase_cmd = SPI_NOR_OCMD_SE;
+		cr2 = 0x02;
+		ret = spi_nor_wrcr2(dev, cr2);
+		data->nor_protocol = PROTO_8D_8D_8D;
+		break;
+	}
+
+	return ret;
+}
+#endif
+
 static int spi_nor_read(const struct device *dev, off_t addr, void *dest,
 			size_t size)
 {
+	struct spi_nor_data *data = dev->data;
```

Comment:
This will trigger a build warning on `data` being unused when `CONFIG_SPI_EXTENDED_MODES` is disabled.

Suggestion: at line 1052:
```patch
-	struct spi_nor_data *data = dev->data;
+	uint8_t read_cmd = SPI_NOR_CMD_READ;
 	const size_t flash_size = dev_flash_size(dev);
 	int ret;
 
+	if (IS_ENABLED(CONFIG_SPI_EXTENDED_MODES))
+		read_cmd = ((struct spi_nor_data *)(dev->data))->read_cmd;
```

Ditto in `spi_nor_write()`.